package/contents/scripts/google_redirect.py

```diff
@@ -0,0 +1,69 @@
+"""Script to handle oauth redirects from Google"""
+
+import json
+import urllib.parse
+import urllib.request
+import urllib.error
+import argparse
+from http.server import BaseHTTPRequestHandler, HTTPServer
+from urllib.parse import urlparse, parse_qs
+
+client_id = client_secret = listen_port = None
+
+
+def exchange_code_for_token(code):
+    # Exchange code for token from https://oauth2.googleapis.com/token
+    # using the following POST request:
+    token_params = {
+        "code": code,
+        "client_id": client_id,
+        "client_secret": client_secret,
+        "redirect_uri": "http://127.0.0.1:{}/".format(listen_port),
+        "grant_type": "authorization_code",
+    }
+    data = urllib.parse.urlencode(token_params).encode("utf-8")
+    req = urllib.request.Request("https://oauth2.googleapis.com/token", data)
+    response = urllib.request.urlopen(req)
+    token_data = json.loads(response.read().decode("utf-8"))
+    return token_data
+
+
+class OAuthRedirectHandler(BaseHTTPRequestHandler):
+    def do_GET(self):
+        query = urlparse(self.path).query
+        params = parse_qs(query)
+        # handle OAuth redirect here
+        if "code" in params:
+            code = params["code"][0]
+            try:
+                token_data = exchange_code_for_token(code)
+            except urllib.error.HTTPError as e:
+                print(e.read().decode("utf-8"))
+                self.wfile.write(b"Handling redirect failed.")
+                raise SystemExit(1)
+            print(json.dumps(token_data, sort_keys=True))
+
+            self.send_response(200)
+            self.send_header("Content-type", "text/html")
+            self.end_headers()
+            self.wfile.write(
+                b"OAuth redirect handled successfully. You can close this tab now."
+            )
+            raise SystemExit(0)
+        self.wfile.write(b"Missing code parameter in redirect.")
```

Comment:
What is the path of this file? Or is it stdout? or stderr? Is there a reason for using write instead of `print()` (no newline)?

https://docs.python.org/3/library/http.server.html#http.server.BaseHTTPRequestHandler.wfile

Reply:
`self.wfile` is to write the response back to the client, in this case the browser (to guide the user what to do). For example, after a successful redirect, the user will see this in their browser:

![Screenshot_20230419_232904](https://user-images.githubusercontent.com/815873/233204339-e5494d97-9e20-4201-bd5b-f4cc7320d583.png)

The print output is for communication with the QML script (using the `executable.exec` helper).